keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,276 @@
+- *Bucket*: a quota pool. The quota will be taken from the bucket when it's
+acquired by any request and will be returned to the bucket when the request
+finishes. A bucket has following attributes:
+  - *Reserved Quota*: an initial guaranteed concurrency can be provided by
+  the bucket
+  - *Shared Quota*: an initial shared concurrency can be provided by the bucket.
+  Shared quota cannot be consumed by requests queued at lower priorities
+  than the bucket's.
+  - *Weight*: only works for shared quota. The higher weight a bucket has,
+  the more chance the requests hitting the buckets can be executed from the
+  queue.
+  - *Priority*: a fixed/preset set of priorities ranged 1...N. The lower
+  number means higher priority. Each priority has one or more buckets providing
+  quotas and buckets from higher priority can "borrow" shared quotas from
+  the lowers.
+- *Extra Shared Quota / ESQ*: global shared quota provided by a logical
+bucket at a virtual "lowest" priority. Requests doesn't match any priority
+will be queued in this logical "lowest" priority until ESQ distributes quota
+to it. ESQ can also be consumed by requests from all priorities. Generally,
+ESQ is a default common concurrency in the system and it's supposed to be
+auto-tunable by a moving average observed by the system. Note that requests
+queued in this extra bandwidth are popped strictly in a FIFO order.
+
+##### Bandwidth Goal
+
+We can define __Shared Bandwidth__ as a goal for priority band __p__, and
+we have N fixed priority:
+
+- *Defined Shared Bandwidth(p) / DSB(p)*: sum of shared quotas provided by
+buckets at Priority p.
+- *System Total Shared Bandwidth / STSB*: sum\[ DSB(i) over 1..N \] + ESQ
+- *Maximum Usable Shared Bandwidth(p) / MUSB*:
+  - if p == 1: STSB
+  - else: STSB - sum\[ STSB - sum\[MUSB(i) over 1...p-1 \]
+
+Generally, the system's overall shared bandwidth consists of a sum of shared
+quotas defined by all buckets and an auto-tunable extra shared bandwidth. In
+this so-called shared bandwidth, a request w/ higher priority can always be
+executed prior to the requests w/ lower priority. Which is, at worst case,
+when requests from top priority completely occupies the shared bandwidth,
+then requests from lower bands can only be executed under its reserved quota.
+And those "wild" request doesn't match any buckets will be rejected.
+
+##### Overview
+
+Within each priority band, we have a set of buckets matching the priority and
+a WRR request queue. Request that missed reserved quota will be queued into
+the WRR queue w/ the weight of its bucket. Note that reserving quota of
+buckets at priority N can also re-claim requests queued at N.
+
+All requests will be pre-processed before it's actually executed by 3 stages:
+
+0. We find a matched bucket by applying BucketBinding API.
+1. *Reserving*: We check if the bucket's reserved quota. If it's sufficient,
+the request will consume one reserved quota from the bucket and be executed
+immediately.
```

Comment:
I don't understand why it is executed immediately. Doesn't this imply that identical requests might either skip a queue or go into a queue just depending on whether another request happened to have just finished?

I don't like this behavior; I'd expect such requests to still go into a queue, which is drained when quota becomes available again.

To say it another way, for a given group of requests, I don't want the latency behavior to be bimodal, with some getting instant service and some waiting for a queue. I strongly prefer to give all requests in the class the same sort of latency, otherwise I think it will be very hard to optimize our tail latency.

Reply:
okay, let's adjust reserved quota logic a bit. now all requests will be queued, and reserved quota from one bucket only feeds its own queue, while shared quota is consumed per priority, higher than the bucket

Reply:
That sounds like a big improvement, thanks!

Reply:
i've already update design (will update this KEP soon too) and also the implementation. i believe they are running well in our test cluster.

all requests will be queued immediately when it hit the rate-limit filter handler. and quotas are continuesly sent to the the drainer to handle the request:

- when receiving reserved quota: the drainer pops pending the request directly from the queue identified w/ the bucket name from the bucket's priority band. or rather, reserved quota from bucket A can only be consumed by the requests exactly matches A regardless of priority.
-

Reply:
Update this doc.



> To say it another way, for a given group of requests, I don't want the latency behavior to be bimodal, with some getting instant service and some waiting for a queue. I strongly prefer to give all requests in the class the same sort of latency, otherwise I think it will be very hard to optimize our tail latency.

This explanation makes the argument against more clear.